src/state/auto-scroller/fluid-scroller.js

```diff
@@ -0,0 +1,329 @@
+};
+
+type WithPlaceholderResult = {|
+  current: Position,
+  max: Position,
+|}
+
+const withPlaceholder = (
+  droppable: DroppableDimension,
+  draggable: DraggableDimension,
+): ?WithPlaceholderResult => {
+  const closest: ?ClosestScrollable = droppable.viewport.closestScrollable;
+
+  if (!closest) {
+    return null;
+  }
+
+  const isOverHome: boolean = droppable.descriptor.id === draggable.descriptor.droppableId;
+  const max: Position = closest.scroll.max;
+  const current: Position = closest.scroll.current;
+
+  // only need to add the buffer for foreign lists
+  if (isOverHome) {
+    return { max, current };
+  }
+
+  const spaceForPlaceholder: Position = patch(
+    droppable.axis.line,
+    draggable.placeholder.withoutMargin[droppable.axis.size]
+  );
+
+  const newMax: Position = add(max, spaceForPlaceholder);
+  // because we are pulling the max forward, on subsequent updates
+  // it is possible for the current position to be greater than the max
+  // as such we need to ensure that the current position is never bigger
+  // than the max position
+  const newCurrent: Position = {
+    x: Math.min(current.x, newMax.x),
+    y: Math.min(current.y, newMax.y),
+  };
+
+  return {
+    max: newMax,
+    current: newCurrent,
+  };
+};
+
+type Api = {|
+  scrollWindow: (offset: Position) => void,
+  scrollDroppable: (id: DroppableId, offset: Position) => void,
+|}
+
+type ResultFn = (state: State) => void;
+type ResultCancel = { cancel: () => void };
+
+export type FluidScroller = ResultFn & ResultCancel;
+
+export default ({
+  scrollWindow,
+  scrollDroppable,
+}: Api): FluidScroller => {
+  const scheduleWindowScroll = rafSchd(scrollWindow);
+  const scheduleDroppableScroll = rafSchd(scrollDroppable);
+
+  const scroller = (state: State): void => {
+    const drag: ?DragState = state.drag;
+    if (!drag) {
+      console.error('Invalid drag state');
+      return;
+    }
+
+    const center: Position = drag.current.page.center;
+
+    // 1. Can we scroll the viewport?
+
+    const draggable: DraggableDimension = state.dimension.draggable[drag.initial.descriptor.id];
+    const subject: Area = draggable.page.withMargin;
+    const viewport: Area = getViewport();
+    const requiredWindowScroll: ?Position = getRequiredScroll({
+      container: viewport,
+      subject,
+      center,
+    });
+
+    if (requiredWindowScroll && canScrollWindow(requiredWindowScroll)) {
+      scheduleWindowScroll(requiredWindowScroll);
+      return;
+    }
+
+    // 2. We are not scrolling the window. Can we scroll a Droppable?
+
+    const droppable: ?DroppableDimension = getBestScrollableDroppable({
+      center,
+      destination: drag.impact.destination,
+      droppables: state.dimension.droppable,
+    });
+
+    // No scrollable targets
+    if (!droppable) {
+      return;
+    }
+
+    // We know this has a closestScrollable
+    const closestScrollable: ClosestScrollable = (droppable.viewport.closestScrollable : any);
```

Comment:
might be able to change line 285 to `if (!droppable || !droppable.viewport.closestScrollable) {` to avoid having to cast through any here?